core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -425,36 +425,35 @@ class Partition(val topicPartition: TopicPartition,
   }
 
   /**
-   * Checks if the topic ID provided in the request is consistent with the topic ID in the log.
+   * Checks if the topic ID received is consistent with the topic ID in the log.
    * If a valid topic ID is provided, and the log exists but has no ID set, set the log ID to be the request ID.
    *
-   * @param requestTopicId the topic ID from the request
+   * @param receivedTopicId the topic ID from the LeaderAndIsr request or from the metadata records
    * @return true if the request topic id is consistent, false otherwise
    */
-  def checkOrSetTopicId(requestTopicId: Uuid): Boolean = {
-    // If the request had an invalid topic ID, then we assume that topic IDs are not supported.
-    // The topic ID was not inconsistent, so return true.
-    // If the log is empty, then we can not say that topic ID is inconsistent, so return true.
-    if (requestTopicId == null || requestTopicId == Uuid.ZERO_UUID)
-      true
-    else {
+  def checkOrSetTopicId(receivedTopicId: Uuid, usingRaft: Boolean): Boolean = {
+    // If the request had an invalid topic ID, then we assume that topic IDs are not supported so ID is consistent.
+    // This is only the case when from LeaderAndIsr Request. Raft code should never have invalid topic IDs.
+    if (receivedTopicId == null || receivedTopicId == Uuid.ZERO_UUID) {
+      if (usingRaft) false else true
```

Comment:
Maybe just `!usingRaft // only okay when not using Raft`

Is this okay when using ZooKeeper with interBrokerProtocolVersion >= KAFKA_2_8_IV0?

Reply:
There are some scenarios during upgrades where we may have invalid topic IDs when interBrokerProtocolVersion >= KAFKA_2_8_IV0 for the receiving broker (ie, the controller may be a lower IBP)